### DISCUSSION TITLE
`unsafe-eval` in development mode

### DISCUSSION THREAD

**Discussion by ryan-codingintrigue** [vercel/next.js#20294] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

As it stands, Next.js 10 forces the user to use Webpack's eval-source-map in development mode. Whilst I understand the speed gains of this decision, the need to override this is absolutely critical.

We do not use `unsafe-eval` in our production code for good reason. We also avoid the use of `eval` in our own codebase through code review. However, if in development we introduce a third-party dependency which *does* use `eval` or `new Function` then we will not be aware of this until the code hits `production`.

This may require a complete rework of functionality and potentially removal of the library.

I think this decision is introducing product risk in favor or developer experience and as a result we have been stuck on 9.5 for a long time.

## Describe the solution you'd like

One of the following:

a) An override flag to turn eval-source-map off at build time. It can still be on by default, but it should not be tied to `NODE_ENV` alone in my opinion.
b) Allow this code to be configurable by config:
https://github.com/vercel/next.js/blob/f06c58911515d980e25c33874c5f18ade5ac99df/packages/next/build/webpack/config/blocks/base.ts#L22
If we are happy enough to allow this for Windows clients, I don't see why we can't allow it for users who wish to explicitly set it.
c) Revert this commit: https://github.com/vercel/next.js/commit/435bf657847329c753b6c0633f6580a62dddb8ba

I do not see why we should be blocking users from making decisions. A warning should be enough, without the revert.

## Describe alternatives you've considered

We have attempted to implement a custom server via Express to override the addition of `unsafe-inline` by the Next.js server, but somewhat frustratingly the Webpack `devtool` flag is ignored, as per this discussion:
https://github.com/vercel/next.js/issues/15823#issuecomment-678755356
And there is no way to reverse this config decision.

**Comment by daithimaolain** [comment]
Yes, this.

**Comment by aidev42** [comment]
Did you ever find a solution? Even when you set your config to `devtools: false` NextJS ignores/overrides this to turn them back on in dev environment!

Has anyone found a solution to getting devtools off? Causing a nightmare for CSP for our secure app.

Maybe we simply can't use NextJS as out frontend?

**Reply by ryan-codingintrigue** [reply]
@aidev42 We forced the use of `inline-source-map` by defining a read-only property on the `config` object in `next.config.js`.

```js
module.exports = {
  webpack: (config, { isServer }) => {
    if (!isServer) {
      Object.defineProperty(config, "devtool", {
        get: () => "inline-source-map",
        set: () => { }
      });
    }
  }
};
```
That prevents the code in the original question from applying the logic.

This worked for our project, however I can't guarantee it won't have a knock-on impact on some of the Next.js developer experience functionality that supposedly requires a different `devtool` value.

**Reply by microhobby** [reply]
I'm getting this now:

```
Warning: Reverting webpack devtool to 'eval-source-map'.
Changing the webpack devtool in development mode will cause severe performance regressions.
Read more: https://nextjs.org/docs/messages/improper-devtool
```

So, I as a developer don't have the choice 🤡
